### PR TITLE
NMS-12788: fully enable NTLMv2; get SMBv2 by move to jcifs-ng and j-interop-ng

### PR DESCRIPTION
**Pull request by dschlenk**
### All Contributors

* [X] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [X] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [X] Have you made a comment in that issue which points back to this PR?
* [X] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [X] If this is a new or updated feature, is there documentation for the new behavior?
* [X] If this is new code, are there unit and/or integration tests?
* [X] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-12788

### CODE REVIEW

**Comment by RangerRick** [comment]
FYI, I ran this as a [full smoke test run](https://app.circleci.com/pipelines/github/OpenNMS/opennms/7555/workflows/66fb7fe2-e042-4853-a5f4-7eda70bc3d57) and it all passed

**Comment by fooker** [comment]
Starting some simple tests without any authentication: The CIFS Monitor works well, but the SMB Monitor crashes with the following Stack Trace:

```
java.lang.ClassCastException: jcifs.netbios.UniAddress cannot be cast to jcifs.netbios.NbtAddress
        at org.opennms.netmgt.poller.monitors.SmbMonitor.poll(SmbMonitor.java:103) [opennms-services-27.0
.0-SNAPSHOT.jar:?]
        at org.opennms.netmgt.poller.client.rpc.PollerClientRpcModule$1.get(PollerClientRpcModule.java:77
) [org.opennms.features.poller.client-rpc-27.0.0-SNAPSHOT.jar:?]
        at org.opennms.netmgt.poller.client.rpc.PollerClientRpcModule$1.get(PollerClientRpcModule.java:71
) [org.opennms.features.poller.client-rpc-27.0.0-SNAPSHOT.jar:?]
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604) [?:1.8.0_2
65]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_265]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_265]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_265]
```

**Comment by mfuhrmann** [comment]
I've tested a Jcifs poller using authentication which worked fine.

Example:

```xml
      <service name="Jcifs" interval="300000" user-defined="true" status="on">
         <parameter key="retry" value="5"/>
         <parameter key="timeout" value="1000"/>
         <parameter key="domain" value="server"/>
         <parameter key="username" value="user"/>
         <parameter key="password" value="password"/>
         <parameter key="path" value="/folder/folder/"/>
         <parameter key="mode" value="folder_not_empty"/>
         <parameter key="folderIgnoreFiles" value="Value1|Value2"/>
      </service>
```

**Comment by dschlenk** [comment]
Thanks @fooker I will check into that SMB Monitor issue. Sorry for the delay, all.

**Comment by dschlenk** [comment]
Okay, new build with SmbMonitor fix passed CI

**Comment by dschlenk** [comment]
OK, I've real-world tested WmiMonitor, SmbMonitor, JCifsMonitor, WMI data collection and all work with the latest build, doing the work directly. I have not done any real world tests with a minion. So remaining testing is:
- [ ] the scenario from [NMS-9763](https://issues.opennms.org/browse/NMS-9763) involving MD4 with the RadiusAuthMonitor, with and without a minion
- [ ] the items above, but with a minion

**Comment by RangerRick** [comment]
@fooker have you had a chance to try out the final minion tests here?

**Comment by mfuhrmann** [comment]
I think I can test https://github.com/OpenNMS/opennms/pull/3134#issuecomment-689809514 with a Minion, too.
@dschlenk I've just had a look at the toc in https://wiki.opennms.org/wiki/WmiConfiguration ... is there an easier/faster way to include WMI collections/monitors without doing all that voodoo?

**Comment by dschlenk** [comment]
@mfuhrmann In short, no, it's very required. Theoretically if we substantially changed how we used j-interop we could avoid needing to change registry permissions on those keys, but I don't have time to go down that road ATM, especially for something that has a valid replacement.

I'm attempting to get a minion deployment up in my lab to test this stuff too.

**Comment by dschlenk** [comment]
Detection and likely monitoring / data collection does not work via minion because 
```
Caused by: java.lang.ClassNotFoundException: jcifs.netbios.NbtAddress not found by wrap_file__opt_minion_.m2_org_jinterop_j-interopdeps_3.0.0_j-interopdeps-3.0.0.jar [411]
```
so I guess I missed the part where I define where that class comes from since that's a jcifs class, not a j-interop class. The jCIFS NG bundle is loaded and includes that class:
```
[schlazor@localhost minion]$ jar -tf data/cache/bundle137/version0.0/bundle.jar  | grep NbtAddress
jcifs/netbios/NbtAddress.class
```
I gotta go watch a soccer game but will look into it more later today hopefully / maybe.

File: container/features/src/main/resources/features.xml [hunk 1]
```diff
@@ -186,7 +189,8 @@
         <bundle>mvn:org.eclipse.persistence/org.eclipse.persistence.antlr/${eclipselinkVersion}</bundle>
     </feature>
     <feature name="jcifs" version="${jcifsVersion}" description="jcifs">
-        <bundle>wrap:mvn:org.samba.jcifs/jcifs/${jcifsVersion}</bundle>
+        <feature>bcprov</feature>
+        <bundle>wrap:mvn:eu.agno3.jcifs/jcifs-ng/${jcifsVersion}</bundle>
```
Comment (cgorantla) [inline container/features/src/main/resources/features.xml:193]: Looks like this is already bundle for it. Don't need to wrap it.

File: container/features/src/main/resources/features.xml [hunk 2]
```diff
@@ -805,7 +809,8 @@
         <feature>commons-cli</feature>
         <feature>opennms-collection-api</feature>
         <feature>opennms-provisioning-api</feature>
-        <bundle dependency="true">wrap:mvn:org.jinterop/j-interop/${jinteropVersion}</bundle>
+        <feature>jcifs</feature>
+        <bundle dependency="true">wrap:mvn:org.jinterop/j-interop/${jinteropVersion}$Import-Package=jcifs,jcifs.smb,jcifs.util,jcifs.netbios,ndr,rpc,rpc.core,rpc.ncacn_np,rpc.pdu,rpc.security.ntlm</bundle>
```
Comment (cgorantla) [inline container/features/src/main/resources/features.xml:813]: Update these bundles with `com.github.skyghis/j-interop-ng`  and `com.github.skyghis/j-interop-ng-deps`

**Comment by dschlenk** [comment]
OK! The WMI Detector works great via Minion. It turns out that neither the WMI nor the JCIFS Monitor have been modified for support on the Minion, so there's nothing to test there. I created [NMS-12942](https://issues.opennms.org/browse/NMS-12942) and [NMS-12943](https://issues.opennms.org/browse/NMS-12943) to track altering them to make that available. Also, while the `WmiCollector` is present on the minion, it does not function in the current release, nor in this branch, as the parts that expect configuration to be present on the filesystem were never updated to work without. I created [NMS-12944](https://issues.opennms.org/browse/NMS-12944) to track that issue.

So, while there are gaps in WMI support via Minion, this branch does not make those gaps wider.

As far as testing the branch goes, just the `RadiusAuthMonitor` remains. I don't have an environment where I can test that.

**Comment by cgorantla** [comment]
@dschlenk Looks like you didn't push updated commits yet.

**Comment by dschlenk** [comment]
whoops! sorry about that. All set now.

File: container/karaf/src/main/filtered-resources/etc/system.properties [hunk 3]
```diff
@@ -177,10 +177,13 @@ karaf.handle.sigterm = false
...
+# @see https://issues.opennms.org/browse/NMS-12788
+
 # Original:
 #org.apache.sshd.security.registrars = \
 #  org.apache.sshd.common.util.security.bouncycastle.BouncyCastleSecurityProviderRegistrar,\
 #  org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderRegistrar
 #
 # Edited:
-#org.apache.sshd.security.registrars = org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderRegistrar
+org.apache.sshd.security.registrars = org.apache.sshd.common.util.security.eddsa.EdDSASecurityProviderRegistrar
```
Comment (cgorantla) [inline container/karaf/src/main/filtered-resources/etc/system.properties:189]: Still need this system property ?

Comment (dschlenk) [reply container/karaf/src/main/filtered-resources/etc/system.properties:189]: Yes - the SSH bundle won't start in the main OpenNMS karaf container without it.

**Review by cgorantla — APPROVED** [review]
LGTM

**Pull request merged**